packages/multichain/src/scope/transform.ts

```diff
@@ -0,0 +1,131 @@
+import type { CaipReference } from '@metamask/utils';
+import { cloneDeep } from 'lodash';
+
+import type {
+  ExternalScopeObject,
+  ExternalScopesObject,
+  ScopeString,
+  ScopeObject,
+  ScopesObject,
+} from './types';
+import { parseScopeString } from './types';
+
+/**
+ * Returns a list of unique items
+ *
+ * @param list - The list of items to filter
+ * @returns A list of unique items
+ */
+export const getUniqueArrayItems = <Value>(list: Value[]): Value[] => {
+  return Array.from(new Set(list));
+};
+
+/**
+ * Normalizes a ScopeString and ExternalScopeObject into a separate
+ * ScopeString and ScopeObject for each reference in the `references`
+ * value if defined and adds an empty `accounts` array if not defined.
+ *
+ * @param scopeString - The string representing the scope
+ * @param externalScopeObject - The object that defines the scope
+ * @returns a map of caipChainId to ScopeObjects
+ */
+export const normalizeScope = (
+  scopeString: string,
+  externalScopeObject: ExternalScopeObject,
+): ScopesObject => {
+  const { references, ...scopeObject } = externalScopeObject;
+  const { namespace, reference } = parseScopeString(scopeString);
+
+  const normalizedScopeObject = {
+    accounts: [],
+    ...scopeObject,
+  };
+
+  // Scope is already a CAIP-2 ID and has no references to flatten
+  if (!namespace || reference || !references) {
+    return { [scopeString]: normalizedScopeObject };
+  }
+
+  const scopeMap: ScopesObject = {};
+  references.forEach((nestedReference: CaipReference) => {
+    scopeMap[`${namespace}:${nestedReference}`] = cloneDeep(
+      normalizedScopeObject,
+    );
+  });
+  return scopeMap;
+};
+
+export const mergeScopeObject = (
+  scopeObjectA: ScopeObject,
+  scopeObjectB: ScopeObject,
+) => {
+  const mergedScopeObject: ScopeObject = {
+    methods: getUniqueArrayItems([
+      ...scopeObjectA.methods,
+      ...scopeObjectB.methods,
+    ]),
+    notifications: getUniqueArrayItems([
+      ...scopeObjectA.notifications,
+      ...scopeObjectB.notifications,
+    ]),
+    accounts: getUniqueArrayItems([
+      ...scopeObjectA.accounts,
+      ...scopeObjectB.accounts,
+    ]),
+  };
+
+  if (scopeObjectA.rpcDocuments || scopeObjectB.rpcDocuments) {
+    mergedScopeObject.rpcDocuments = getUniqueArrayItems([
+      ...(scopeObjectA.rpcDocuments ?? []),
+      ...(scopeObjectB.rpcDocuments ?? []),
+    ]);
+  }
+
+  if (scopeObjectA.rpcEndpoints || scopeObjectB.rpcEndpoints) {
+    mergedScopeObject.rpcEndpoints = getUniqueArrayItems([
+      ...(scopeObjectA.rpcEndpoints ?? []),
+      ...(scopeObjectB.rpcEndpoints ?? []),
+    ]);
+  }
+
+  return mergedScopeObject;
+};
+
+export const mergeScopes = (
```

Comment:
Our type names definitely could use some work. Perhaps we can hop on a call and work through them together

Reply:
We could if you have time! I have a couple of suggestions in the meantime:

- `KeyedScopeObject` (matches the wording in CAIP-15, but might still be confusing) — then it would be `mergeKeyedScopeObjects` vs. `mergeScopeObjects`
- `ScopeObjectMap` (quite literal, doesn't match the wording in CAIP-15, but seems different enough from `scopeObject` to not cause confusion) — then it would be `mergeScopeObjectMaps` vs. `mergeScopeObjects`.

Do either of those work better?